File: CODE/Treccanimod.f90
!=========================================
!
! FUNZIONE DI TRECCANI
! 
!  X(1) .GE. -2.5  ;  X(2) .LE.  2.5
!
!=========================================

SUBROUTINE FUNCT(N,X,F)

      IMPLICIT NONE

      INTEGER          :: N
      DOUBLE PRECISION :: X(N), F, xaux(n)

      DOUBLE PRECISION :: X1Q

      Xaux(1) = X(1) + 0.25d0*dsqrt(2.d0)
      Xaux(2) = X(2) + 0.25d0*dsqrt(2.d0)
	
      F = 0.D0
      X1Q = Xaux(1)*Xaux(1)
      F = X1Q*X1Q+4.D0*X1Q*Xaux(1)+4.D0*X1Q+Xaux(2)*Xaux(2)

      RETURN

END


subroutine grad(n,x,g)

integer				::n
double precision	::x(n),g(n), Xaux(n)

Xaux(1) = X(1) + 0.25d0*dsqrt(2.d0)
Xaux(2) = X(2) + 0.25d0*dsqrt(2.d0)


g(1) = 4.0d0*Xaux(1)**3+12.0d0*Xaux(1)**2+8.0d0*Xaux(1)

g(2) = 2.0d0*Xaux(2)

return

end subroutine grad

File: Python/treccanimod.py
#=========================================
#
# FUNZIONE DI TRECCANI
# 
#  X(1) .GE. -2.5  ;  X(2) .LE.  2.5
#
#=========================================

import numpy as np

def funct(x):
    x_aux = x + 0.25 * np.sqrt(2.0)
    x1q = x_aux[0] * x_aux[0]
    result = x1q * x1q + 4.0 * x1q * x_aux[0] + 4.0 * x1q + x_aux[1] * x_aux[1]
    return result

def grad(x):
    x_aux = x + 0.25 * np.sqrt(2.0)

    g1 = 4.0 * x_aux[0]**3 + 12.0 * x_aux[0]**2 + 8.0 * x_aux[0]
    g2 = 2.0 * x_aux[1]

    return np.array([g1, g2])

if __name__ == "__main__":
    # Example Usage
    n_value = 2
    x_values = np.array([1.0, 2.0])  # Replace with your actual input values

    result_funct = funct(x_values)
    result_grad = grad(x_values)

    print("Result (Function):", result_funct)
    print("Result (Gradient):", result_grad)
